### ISSUE TITLE
image::open(jpg) - no data found

### ISSUE THREAD

**Issue by elbaro** [image-rs/image#738]
Code:
```
let _ = image::open("...jpg").unwrap();
```

Error:
```
FormatError("no data found")
```

Image:
https://farm1.staticflickr.com/629/33472552930_58c1e6ea1e_o_d.jpg

**Comment by gasabr** [comment]
I tried to fix this and looks like panic is causes by this [line](https://github.com/kaksmet/jpeg-decoder/blob/aaf05e502e8360f8f094e39c2474af320b795f83/src/decoder.rs#L329) in jpeg_decoder crate, so I wrote to the author of the crate, hope he will provide some insights on the issue

**Comment by kaksmet** [comment]
@gasabr Thanks for taking a look at this issue.

The image is a progressive jpeg using both spectral selection and successive approximation. I think the problem occurs because the image does not have a scan that provides the least significant bit for some of the first component's coefficients. Therefore [this check](https://github.com/kaksmet/jpeg-decoder/blob/755d386aa56050b3f0980851c62d35b5e1f4a72e/src/decoder.rs#L218) used for detecting the final scan in the image never passes and no IDCT is performed.

**Comment by gasabr** [comment]
@kaksmet as far as I understood from your comment, the error was caused by the jpeg image with invalid structure, if so, should not the issue be closed?

**Comment by kaksmet** [comment]
Technically the image is incomplete, but I'm not sure the JPEG spec actually requires images with successive approximation to be fully refined.
Since libjpeg(-turbo) successfully decodes this image, I think we should aim for `image`/`jpeg-decoder` to also be able to do it.

@elbaro Do you have any knowledge of what software/hardware produced this image?

**Comment by elbaro** [comment]
No, it's a random image from flikr.
All I know is my browser and other viewers can open it.

**Comment by elbaro** [comment]
```
xxx.jpg Format error: spectral selection is not allowed in non-progressive scan
xxx.jpg Format error: 4 components without Adobe APP14 metadata to tell color space
xxx.jpg failed to fill whole buffer
xxx.jpg Format error: no data found
```

I have dozens of other jpeg samples like above. They have no problem when opened in many applications (web browsers, file explorers, etc).

Their original flikr urls are lost, and I am not sure if it is allowed to upload them to my cloud and post the link here. If someone leaves the email address here I will send the link.

**Comment by kaksmet** [comment]
@elbaro I'm interested in taking a look at the images, you can find the email in my profile.

**Comment by kaksmet** [comment]
I have analyzed the images @elbaro sent to me.

- `Format error: 4 components without Adobe APP14 metadata to tell color space`
This is a bug in `jpeg-decoder`, easy fix.

- `Format error: spectral selection is not allowed in non-progressive scan`
These images violates the JPEG spec, but `libjpeg(-turbo)` allows them (although with a warning). I will change `jpeg-decoder`'s behavior to match `libjpeg`.

- `failed to fill whole buffer`
This image is corrupt, and `jpeg-decoder` currently lacks any kind of error recovery so no image can be produced.

- `Format error: no data found`
The images that produce this error all seem to lack a final scan, like the original image in this bug report.

**Comment by flymonkey00** [comment]
I also encountered this problem, I can open all pictures in mac preview, but can't decode them using this library, so how can I simply fix it or just ignore this error? Thanks a lot.

**Comment by flymonkey00** [comment]
I found that almost images downloaded from Instagram can't be decoded by this library and the error message always be "no data found".

**Comment by HeroicKatora** [comment]
Since the implementation here [just proxies](https://github.com/image-rs/image/blob/master/src/jpeg/decoder.rs) the [jpeg-decoder](https://crates.io/crates/jpeg-decoder) crate, the issue will make more progress if it is reported and evaluated for that crate first. Specifically, the error does not originate within `image` but simply a converted version of the error type of `jpeg-decoder`.

**Comment by HeroicKatora** [comment]
So, if you find that this also applies to Instagram picture then please contribute to the referenced issue in jpeg-decoder (https://github.com/kaksmet/jpeg-decoder/issues/96).

**Comment by HeroicKatora** [comment]
This was fixed with `jpeg-decoder` `v0.1.22`.
